core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -920,36 +929,58 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
       ensureFullPathAndUpdateCache(inodePath);
       inode = inodePath.getInode();
       auditContext.setSrcInode(inode);
-
       List<FileInfo> ret = new ArrayList<>();
-      if (inode.isDirectory()) {
-        try (TempInodePathForDescendant tempInodePath = new TempInodePathForDescendant(inodePath)) {
-          try {
-            mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, inodePath);
-          } catch (AccessControlException e) {
-            auditContext.setAllowed(false);
-            throw e;
-          }
-          for (Inode<?> child : ((InodeDirectory) inode).getChildren()) {
-            child.lockReadAndCheckParent(inode);
-            try {
-              // the path to child for getPath should already be locked.
-              tempInodePath.setDescendant(child, mInodeTree.getPath(child));
-              ret.add(getFileInfoInternal(tempInodePath));
-            } finally {
-              child.unlockRead();
-            }
-          }
-        }
-      } else {
-        ret.add(getFileInfoInternal(inodePath));
+      DescendantType descendantTypeForListStatus =  (listStatusOptions.isRecursive())
+          ? DescendantType.ALL : DescendantType.ONE;
+      listStatusInternal(inodePath, auditContext,
+          descendantTypeForListStatus, ret);
+
+      // If we are listing the status of a directory, we remove the directory info that we inserted
+      if (inode.isDirectory() && ret.size() >= 1) {
+        ret.remove(ret.size() - 1);
       }
+
       auditContext.setSucceeded(true);
       Metrics.FILE_INFOS_GOT.inc();
       return ret;
     }
   }
 
+  private void listStatusInternal(LockedInodePath currInodePath,
+                                  AuditContext auditContext,
+                                  DescendantType descendantType,
+                                  List<FileInfo> statusList)
+      throws FileDoesNotExistException, UnavailableException, AccessControlException {
+    Inode<?> inode = currInodePath.getInode();
+    if (inode.isDirectory()) {
+      try {
+        mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, currInodePath);
+      } catch (AccessControlException e) {
+        auditContext.setAllowed(false);
+        if (descendantType == DescendantType.ALL) {
+          return;
+        } else {
+          throw e;
+        }
+      }
+      if (descendantType == DescendantType.ALL || descendantType == DescendantType.ONE) {
+        DescendantType nextDescendantType = (descendantType == DescendantType.ALL)
+            ? DescendantType.ALL : DescendantType.NONE;
+        for (Inode<?> child : ((InodeDirectory) inode).getChildren()) {
+          try (LockedInodePath childInodePath = mInodeTree.lockFullInodePath(child.getId(),
+              InodeTree.LockMode.READ)) {
+            listStatusInternal(childInodePath, auditContext,
+                nextDescendantType, statusList);
```

Comment:
NIT: will this fit on the previous line?

Reply:
done